From dd36b3b64f303c4a742e4d37eabff26039621e16 Mon Sep 17 00:00:00 2001
From: Sai Vemuri <svemuri@redhat.com>
Date: Tue, 16 Dec 2014 22:22:12 -0500
Subject: [scsi] cxgb4i: additional types of negative advice

Message-id: <1418768535-2637-5-git-send-email-svemuri@redhat.com>
Patchwork-id: 102109
O-Subject: [RHEL7.1 PATCH BZ 1174982 4/7] cxgb4i: additional types of negative advice
Bugzilla: 1174982
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

Upstream : https://git.kernel.org/cgit/linux/kernel/git/davem/net-next.git/commit/?id=928567ada48b772505e5245267d616c7af97edf0

Treat both CPL_ERR_KEEPALV_NEG_ADVICE and CPL_ERR_PERSIST_NEG_ADVICE as
negative advice.

Signed-off-by: Karen Xie <kxie@chelsio.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit 928567ada48b772505e5245267d616c7af97edf0)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/cxgbi/cxgb4i/cxgb4i.c b/drivers/scsi/cxgbi/cxgb4i/cxgb4i.c
index ec54592..34205ca 100644
--- a/drivers/scsi/cxgbi/cxgb4i/cxgb4i.c
+++ b/drivers/scsi/cxgbi/cxgb4i/cxgb4i.c
@@ -846,6 +846,13 @@ static void csk_act_open_retry_timer(unsigned long data)
 
 }
 
+static inline bool is_neg_adv(unsigned int status)
+{
+ return status == CPL_ERR_RTX_NEG_ADVICE ||
+  status == CPL_ERR_KEEPALV_NEG_ADVICE ||
+  status == CPL_ERR_PERSIST_NEG_ADVICE;
+}
+
 static void do_act_open_rpl(struct cxgbi_device *cdev, struct sk_buff *skb)
 {
  struct cxgbi_sock *csk;
@@ -867,7 +874,7 @@ static void do_act_open_rpl(struct cxgbi_device *cdev, struct sk_buff *skb)
          "csk 0x%p,%u,0x%lx. ", (&csk->saddr), (&csk->daddr),
          atid, tid, status, csk, csk->state, csk->flags);
 
- if (status == CPL_ERR_RTX_NEG_ADVICE)
+ if (is_neg_adv(status))
   goto rel_skb;
 
  module_put(THIS_MODULE);
@@ -973,8 +980,7 @@ static void do_abort_req_rss(struct cxgbi_device *cdev, struct sk_buff *skb)
          (&csk->saddr), (&csk->daddr),
          csk, csk->state, csk->flags, csk->tid, req->status);
 
- if (req->status == CPL_ERR_RTX_NEG_ADVICE ||
-     req->status == CPL_ERR_PERSIST_NEG_ADVICE)
+ if (is_neg_adv(req->status))
   goto rel_skb;
 
  cxgbi_sock_get(csk);
-- 
1.7.1